src/event.rs

```diff
@@ -0,0 +1,629 @@
+						}
+					}
+					Err(_err) => {
+						self.channel_manager
+							.force_close_without_broadcasting_txn(
+								&temporary_channel_id,
+								&counterparty_node_id,
+							)
+							.expect(
+								"Failed to force close channel after funding generation failed",
+							);
+					}
+				}
+			}
+			LdkEvent::PaymentClaimable {
+				payment_hash,
+				purpose,
+				amount_msat,
+				receiver_node_id: _,
+				via_channel_id: _,
+				via_user_channel_id: _,
+			} => {
+				log_info!(
+					self.logger,
+					"Received payment from payment hash {} of {} millisatoshis",
+					hex_utils::to_string(&payment_hash.0),
+					amount_msat,
+				);
+				let payment_preimage = match purpose {
+					PaymentPurpose::InvoicePayment { payment_preimage, payment_secret } => {
+						if payment_preimage.is_some() {
+							payment_preimage
+						} else {
+							self.channel_manager
+								.get_payment_preimage(payment_hash, payment_secret)
+								.ok()
+						}
+					}
+					PaymentPurpose::SpontaneousPayment(preimage) => Some(preimage),
+				};
+
+				if let Some(preimage) = payment_preimage {
+					self.channel_manager.claim_funds(preimage);
+				} else {
+					log_error!(
+						self.logger,
+						"Failed to claim payment with hash {}: preimage unknown.",
+						hex_utils::to_string(&payment_hash.0),
+					);
+				}
+			}
+			LdkEvent::PaymentClaimed {
+				payment_hash,
+				purpose,
+				amount_msat,
+				receiver_node_id: _,
+			} => {
+				log_info!(
+					self.logger,
+					"Claimed payment from payment hash {} of {} millisatoshis.",
+					hex_utils::to_string(&payment_hash.0),
+					amount_msat,
+				);
+				let (payment_preimage, payment_secret) = match purpose {
+					PaymentPurpose::InvoicePayment { payment_preimage, payment_secret, .. } => {
+						(payment_preimage, Some(payment_secret))
+					}
+					PaymentPurpose::SpontaneousPayment(preimage) => (Some(preimage), None),
+				};
+				let mut payments = self.inbound_payments.lock().unwrap();
+				match payments.entry(payment_hash) {
+					hash_map::Entry::Occupied(mut e) => {
+						let payment = e.get_mut();
+						payment.status = PaymentStatus::Succeeded;
+						payment.preimage = payment_preimage;
+						payment.secret = payment_secret;
+					}
+					hash_map::Entry::Vacant(e) => {
+						e.insert(PaymentInfo {
+							preimage: payment_preimage,
+							secret: payment_secret,
+							status: PaymentStatus::Succeeded,
+							amount_msat: Some(amount_msat),
+						});
+					}
+				}
+				self.event_queue
+					.add_event(Event::PaymentReceived { payment_hash, amount_msat })
+					.expect("Failed to push to event queue");
+			}
+			LdkEvent::PaymentSent { payment_preimage, payment_hash, fee_paid_msat, .. } => {
+				let mut payments = self.outbound_payments.lock().unwrap();
+				for (hash, payment) in payments.iter_mut() {
+					if *hash == payment_hash {
+						payment.preimage = Some(payment_preimage);
+						payment.status = PaymentStatus::Succeeded;
```

Comment:
I'd consider adding methods to `PaymentInfoStorage` or `PaymentInfo` such that logic for updating the entries is kept in a single place. Likewise elsewhere.

Reply:
Yes, this will happen in #13.